### PR TITLE
postgis location implemented in group and event

### PR DESCRIPTION
**Pull request by codextech**

### CODE REVIEW

File: src/event/event.service.ts [hunk 1]
```diff
@@ -274,9 +297,26 @@ export class EventService {
...
+      }
+
+      // Default radius to 5 kilometers if not provided
+      const searchRadius = radius ?? 5;
+
+      // Find events within the radius using ST_DWithin
+      eventQuery.andWhere(
+        `ST_DWithin(
+          event.locationPoint,
+          ST_SetSRID(ST_MakePoint(:lon, :lat), 4326),
```
Comment (tompscanlan) [inline src/event/event.service.ts:315]: I see 4326 used in a couple of places.  What is that? Can we make a constant?

File: src/database/migrations/1732279680865-LocationColumnInEventAndGroup.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+export class AddLocationPointToEventAndGroup1732279680865
+  implements MigrationInterface
+{
+  name = 'AddLocationPointToEventAndGroup1732279680865';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    const schema = queryRunner.connection.options.name || 'public';
+    // Adding locationPoint column to "groups"
+    await queryRunner.query(
+      `ALTER TABLE "${schema}"."groups" ADD "locationPoint" geography(Point,4326)`,
```
Comment (stanislavkhatko) [inline src/database/migrations/1732279680865-LocationColumnInEventAndGroup.ts:12]: Should there also be an index for that field? Please place in BaseTables instead.

File: src/database/seeds/relational/event/event-seed.service.ts [hunk 3]
```diff
@@ -66,7 +66,7 @@ export class EventSeedService {
         startDate: eventData.startDate,
         endDate: eventData.endDate,
         maxAttendees: eventData.maxAttendees,
-        location: eventData.location ?? '',
```
Comment (stanislavkhatko) [inline src/database/seeds/relational/event/event-seed.service.ts:69]: Location is still a 'string' with address name.

File: src/event/dto/create-event.dto.ts [hunk 4]
```diff
@@ -64,13 +64,6 @@ export class CreateEventDto {
   @IsString()
   type: string;
 
-  @ApiProperty({
-    description: 'The location of the event',
-  })
-  @IsOptional()
-  @IsString()
-  location: string;
```
Comment (stanislavkhatko) [inline src/event/dto/create-event.dto.ts:72]: We still need location string

File: src/event/events.controller.spect.ts [hunk 5]
```diff
@@ -36,7 +36,6 @@ const createEventDto: CreateEventDto = {
   endDate: new Date(),
   group: mockGroup.id,
   type: 'in_person',
-  location: 'Test Location',
```
Comment (stanislavkhatko) [inline src/event/events.controller.spect.ts:39]: bring back locations on event and group

File: src/group/group.service.ts [hunk 6]
```diff
@@ -286,9 +301,26 @@ export class GroupService {
...
+      const [lon, lat] = location.split(',').map(Number);
+      if (isNaN(lon) || isNaN(lat)) {
+        throw new BadRequestException(
+          'Invalid location format. Expected "lon,lat".',
+        );
+      }
+
+      // Default radius to 5 kilometers if not provided
+      const searchRadius = radius ?? 5;
+      console.log(lon, lat, radius);
```
Comment (stanislavkhatko) [inline src/group/group.service.ts:314]: remove log

File: test/utils/functions.ts [hunk 7]
```diff
@@ -61,7 +61,7 @@ export async function createGroupsAndEvents(
     startDate: new Date(),
     endDate: new Date(new Date().getTime() + 24 * 60 * 60 * 1000), // One day after the start date
     maxAttendees: 100,
-    location: 'Virtual',
+    // location: 'Virtual',
```
Comment (stanislavkhatko) [inline test/utils/functions.ts:64]: revert location back

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]
Bring back 'location'. Add index for searchable fields.

File: env-example-relational [hunk 8]
```diff
@@ -91,3 +91,5 @@ TESTING_TENANT_ID=testing
 ZULIP_USERNAME=
 ZULIP_API_KEY=
 ZULIP_REALM=
+
+POSTGIS_SRID=4326
```
Comment (stanislavkhatko) [inline env-example-relational:95]: It shouldn't be here but in `constraints` instead.

File: src/event/event.service.ts [hunk 9]
```diff
@@ -247,16 +262,24 @@ export class EventService {
...
 
     const eventQuery = this.eventRepository
       .createQueryBuilder('event')
       .leftJoinAndSelect('event.user', 'user')
       .leftJoinAndSelect('event.categories', 'categories')
       .leftJoinAndSelect('event.group', 'group')
-      .leftJoinAndSelect('event.attendees', 'attendees')
-      .where('event.status = :status', { status: EventStatus.Published });
+      .leftJoinAndSelect('event.attendees', 'attendees');
+    // .where('event.status = :status', { status: EventStatus.Published });
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:282]: @codextech We do need to filter events by Published, was it for testing?

File: src/event/event.service.ts [hunk 10]
```diff
@@ -274,9 +297,26 @@ export class EventService {
...
+      }
+
+      // Default radius to 5 kilometers if not provided
+      const searchRadius = radius ?? 5;
+
+      // Find events within the radius using ST_DWithin
+      eventQuery.andWhere(
+        `ST_DWithin(
+          event.locationPoint,
+          ST_SetSRID(ST_MakePoint(:lon, :lat), ${process.env.POSTGIS_SRID}),
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:315]: Take SRID from constraints instead of process.env

File: src/event/events.controller.spect.ts (hunk 5 above)
Comment (stanislavkhatko) [inline src/event/events.controller.spect.ts:39]: Why location removed?

File: src/group/group.service.ts [hunk 11]
```diff
@@ -286,9 +301,25 @@ export class GroupService {
...
+        );
+      }
+
+      // Default radius to 5 kilometers if not provided
+      const searchRadius = radius ?? 5;
+      // Find events within the radius using ST_DWithin
+      groupQuery.andWhere(
+        `ST_DWithin(
+          group.locationPoint,
+          ST_SetSRID(ST_MakePoint(:lon, :lat), ${process.env.POSTGIS_SRID}),
```
Comment (stanislavkhatko) [inline src/group/group.service.ts:318]: SRID not env constant

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]
Need to bring back `location` field on event and group as it's a string to store location name.
Missing db index for location
Move SRID to constraints.

File: src/database/seeds/relational/event/event-seed.service.ts (hunk 3 above)
Comment (codextech) [reply src/database/seeds/relational/event/event-seed.service.ts]: i did not get your point here

Comment (stanislavkhatko) [reply src/database/seeds/relational/event/event-seed.service.ts]: Bring back location. What was the reason to remove it?

File: src/group/infrastructure/persistence/relational/entities/group.entity.ts [hunk 12]
```diff
@@ -70,7 +71,17 @@ export class GroupEntity extends EntityRelationalHelper {
   @Column({ type: 'boolean', default: true })
   requireApproval: boolean;
 
+  @Column({
+    type: 'geography',
+    spatialFeatureType: 'Point',
+    srid: 4326,
+    nullable: true,
+  })
+  @Index()
```
Comment (stanislavkhatko) [inline src/group/infrastructure/persistence/relational/entities/group.entity.ts:80]: Need db index as well

File: src/group/infrastructure/persistence/relational/entities/group.entity.ts [hunk 13]
```diff
@@ -70,7 +71,17 @@ export class GroupEntity extends EntityRelationalHelper {
...
+  locationPoint: string;
+
   @Column({ type: 'varchar', length: 255, nullable: true })
+  @Index()
```
Comment (stanislavkhatko) [inline src/group/infrastructure/persistence/relational/entities/group.entity.ts:84]: This index might be usefull, required db index in migration as well

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]

File: src/event/infrastructure/persistence/relational/entities/event.entity.ts [hunk 14]
```diff
@@ -91,7 +91,17 @@ export class EventEntity extends EntityRelationalHelper {
   @Column({ type: 'boolean', default: false })
   requireGroupMembership: boolean;
 
+  @Column({
+    type: 'geography',
+    spatialFeatureType: 'Point',
+    srid: 4326,
```
Comment (tompscanlan) [inline src/event/infrastructure/persistence/relational/entities/event.entity.ts:97]: reference the constant

File: src/database/migrations/1728637873969-BaseTables.ts [hunk 15]
```diff
@@ -89,10 +86,7 @@ export class BaseTables1728637873969 implements MigrationInterface {
...
-    await queryRunner.query(
-      `CREATE INDEX "IDX_${schema}_events_slug" ON "${schema}"."events" ("slug")`,
+      `CREATE TABLE "${schema}"."events" ("createdAt" TIMESTAMP NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP NOT NULL DEFAULT now(), "id" SERIAL NOT NULL, "ulid" character varying, "zulipChannelId" integer, "name" character varying(255) NOT NULL, "slug" character varying(255) NOT NULL, "type" "${schema}"."events_type_enum" NOT NULL, "locationOnline" character varying(255), "description" text NOT NULL, "startDate" TIMESTAMP NOT NULL, "endDate" TIMESTAMP, "maxAttendees" integer, "requireApproval" boolean NOT NULL DEFAULT false, "approvalQuestion" character varying(255), "requireGroupMembership" boolean NOT NULL DEFAULT false, "location" character varying(255), "lat" double precision, "lon" double precision, "status" "${schema}"."events_status_enum", "visibility" "${schema}"."events_visibility_enum", "allowWaitlist" boolean NOT NULL DEFAULT false, "imageId" integer, "userId" integer, "groupId" integer, "locationPoint" geography(Point, 4326), CONSTRAINT "REL_35515e57a42f4fd00a4172371b" UNIQUE ("imageId"), CONSTRAINT "PK_40731c7151fe4be3116e45ddf73" PRIMARY KEY ("id"))`,
```
Comment (tompscanlan) [inline src/database/migrations/1728637873969-BaseTables.ts:89]: reference the constant

File: src/database/migrations/1728637873969-BaseTables.ts [hunk 16]
```diff
@@ -69,10 +69,7 @@ export class BaseTables1728637873969 implements MigrationInterface {
...
-    );
-    await queryRunner.query(
-      `CREATE INDEX "IDX_${schema}_groups_slug" ON "${schema}"."groups" ("slug")`,
+      `CREATE TABLE "${schema}"."groups" ("createdAt" TIMESTAMP NOT NULL DEFAULT now(), "updatedAt" TIMESTAMP NOT NULL DEFAULT now(), "id" SERIAL NOT NULL,"ulid" character varying(26) NOT NULL, "zulipChannelId" integer, "name" character varying(255) NOT NULL, "slug" character varying(255) NOT NULL, "description" text NOT NULL, "status" "${schema}"."groups_status_enum", "visibility" "${schema}"."groups_visibility_enum", "requireApproval" boolean NOT NULL DEFAULT true, "location" character varying(255), "lat" double precision, "lon" double precision, "createdById" integer, "imageId" integer, "locationPoint" geography(Point, 4326), CONSTRAINT "REL_44626591821828ce1d26311312" UNIQUE ("imageId"), CONSTRAINT "PK_659d1483316afb28afd3a90646e" PRIMARY KEY ("id"))`,
```
Comment (tompscanlan) [inline src/database/migrations/1728637873969-BaseTables.ts:72]: reference the constant

File: src/database/seeds/relational/event/event-seed.service.ts (hunk 3 above)
Comment (tompscanlan) [reply src/database/seeds/relational/event/event-seed.service.ts]: This isn't addressed.  Stan needs the location

File: src/database/migrations/1732279680865-LocationColumnInEventAndGroup.ts (hunk 2 above)
Comment (codextech) [reply src/database/migrations/1732279680865-LocationColumnInEventAndGroup.ts]: its already done.

File: src/event/event.service.ts [hunk 17]
```diff
@@ -221,9 +245,26 @@ export class EventService {
     }
 
     if (location) {
-      eventQuery.andWhere('event.location LIKE :location', {
-        location: `%${location}%`,
-      });
+      // Split and validate location
+      const [lon, lat] = location.split(',').map(Number);
```
Comment (stanislavkhatko) [inline src/event/event.service.ts:249]: from query you'll receive:
```javascript
location: string,
lat: number,
lon: number
```

**Review by stanislavkhatko — CHANGES_REQUESTED** [review]
Query location fix

**Pull request merged**